### PR TITLE
add OctavePackage

### PR DESCRIPTION
**Pull request by davydden**
EDIT: I think I figure out the installation. 

Overriding `uninstall` is still tricky as I can't get to `octave` executable the same way I did with `install`:
```
    def do_uninstall(self, force=False):
        """Uninstall this package by spec."""
        name = 'spline'
        inspect.getmodule(self).octave('pkg uninstall %s' % name)
        # delegate to instance-less method.
        Package.uninstall_by_spec(self.spec, force)
```
produces
```
==> Error: 'module' object has no attribute 'octave'
```

I guess I leave this for another time unless someone knows an easy fix.

----

~~it's WIP as I am bit confused why `setup_dependent_package()` is not called, defined in `octave` package~~ https://github.com/spack/spack/blob/develop/var/spack/repos/builtin/packages/octave/package.py#L234-L241

Maybe I just need to follow `PythonPackage` which has
```
import inspect
def python(self, *args, **kwargs):
        inspect.getmodule(self).python(*args, **kwargs)
```
and if I understand correctly is being used as `self.python`.

### CODE REVIEW

**Comment by adamjstewart** [comment]
You also need to edit `lib/spack/docs/packaging_guide.rst` and `lib/spack/spack/cmd/create.py`. See #4936 for the last build system I added.

**Comment by davydden** [comment]
thanks @adamjstewart .

> lib/spack/spack/cmd/create.py. 

apparently, we already have `OctavePackageTemplate`. I will adjust it...

File: lib/spack/spack/build_systems/octave.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+    extends('octave')
+    depends_on('octave', type=('build', 'run'))
+
+    def install(self, spec, prefix):
+        """Install the package from the archive file"""
+
+        # octave does not like those environment variables to be set:
+        os.environ.pop('CC', '')
+        os.environ.pop('CXX', '')
+        os.environ.pop('FC', '')
```
Comment (adamjstewart) [inline lib/spack/spack/build_systems/octave.py]: I'm wondering if it would be better to separate this out into `setup_environment`.

Comment (davydden) [reply lib/spack/spack/build_systems/octave.py]: done.

Comment (davydden) [reply lib/spack/spack/build_systems/octave.py]: minor inconvenience is that we have warnings:
```
==> Warning: Suspicious requests to set or unset 'CC' found
==> Warning: 	    	env.set('CC', join_path(link_dir, compiler.link_paths['cc'])) at /Users/davydden/spack/lib/spack/spack/build_environment.py:146
==> Warning: 	--->	spack_env.unset('CC') at /Users/davydden/spack/lib/spack/spack/build_systems/octave.py:54
==> Warning: Suspicious requests to set or unset 'CXX' found
==> Warning: 	    	env.set('CXX', join_path(link_dir, compiler.link_paths['cxx'])) at /Users/davydden/spack/lib/spack/spack/build_environment.py:149
==> Warning: 	--->	spack_env.unset('CXX') at /Users/davydden/spack/lib/spack/spack/build_systems/octave.py:55
==> Warning: Suspicious requests to set or unset 'FC' found
==> Warning: 	    	env.set('FC', join_path(link_dir, compiler.link_paths['fc'])) at /Users/davydden/spack/lib/spack/spack/build_environment.py:155
==> Warning: 	--->	spack_env.unset('FC') at /Users/davydden/spack/lib/spack/spack/build_systems/octave.py:56
```
but I am ok with it.

**Pull request merged**